File: src/bsp.zig
const std = @import("std");
const config = @import("config");

const use_rpi3 = std.mem.eql(u8, config.board, "bsp_rpi3");
const use_rpi4 = std.mem.eql(u8, config.board, "bsp_rpi4");

pub const board = if (use_rpi3 or use_rpi4) @import("bsp/raspberrypi.zig") else null;
pub const driver = if (use_rpi3 or use_rpi4) @import("bsp/device_driver.zig") else null;

File: src/kernel/bsp/bsp.cpp
#include "bsp.hpp"

#if BOARD == bsp_rpi3 || BOARD == bsp_rpi4
#include "raspberrypi/raspberrypi.hpp"
#endif

namespace BSP {

void Board::init() {
#if BOARD == bsp_rpi3 || BOARD == bsp_rpi4
  Driver::BSP::RaspberryPi::RaspberryPi::init();
#else
#error Unknown board
#endif
}

size_t Board::getDefaultLoadAddr() {
#if BOARD == bsp_rpi3 || BOARD == bsp_rpi4
  return Driver::BSP::RaspberryPi::RaspberryPi::getDefaultLoadAddr();
#else
#error Unknown board
#endif
}

} // namespace BSP